File: rice_cooker.lua
local RICE_STORAGE = 0
local WATER_STORAGE = 0
local ingredients_basket = {}
local TIMER = 0
local keep_warm_mode = false
local power_on = false
local lid_closed = true
local is_heating = false
local is_under_pressure = false
local is_warm = false

function DisplayOptions()
    print("Options:")
    print("1 - Plug In/Plug Out")
    print("2 - Open/Close")
    print("3 - Keep warm mode ")
    print("4 - Check status light")
    print("5 - Put ingredients in the main storage")
    print("6 - Put ingredients in the steaming basket")
    print("7 - Set the TIMER")
    print("8 - Stop the TIMER")
    print("9 - Exit")
end

function MainMenu()
    DisplayOptions()
    local choice = io.read("*n")

    local success,result = pcall(function()
        if choice == 1 then
            power_on = not power_on
            print("Rice cooker is now " .. (power_on and "plugged in" or "plugged out"))
            MainMenu()
        elseif choice == 2 then
            lid_closed = not lid_closed
            print("Rice cooker is now " .. (lid_closed and "closed" or "open"))
            MainMenu()
        elseif choice == 3 then
            keep_warm_mode = not keep_warm_mode
            print("Keep warm mode is now " .. (keep_warm_mode and "on" or "off"))
            MainMenu()
        elseif choice == 4 then
            if is_heating == true then
                print("Rice cooker is heating")
            elseif is_under_pressure == true then
                print("Rice cooker is under pressure")
            elseif is_warm == true then
                print("Rice cooker is warming up")
            else
                print("Rice cooker is not plugged in")
            end
            print("Status light is " .. (power_on and "on" or "off"))
            MainMenu()
        elseif choice == 5 then
            MainStorageMenu()
        elseif choice == 6 then
            SteamingBasketMenu()
        elseif choice == 7 then
            SetTIMER()
            MainMenu()
        elseif choice == 8 then
            StopTIMER()
            MainMenu()
        elseif choice == 9 then
            print("Exiting the rice cooker.")
        else
            error("Invalid choice. Please try again.")
            MainMenu()
        end
    end)
    if not success then
        print("Error:",result)
        MainMenu()
    end
end

function MainStorageMenu()
    print("Main Storage Options:")
    print("1 - Put rice (max 10 kg)")
    print("2 - Remove rice")
    print("3 - Put water (max 10 L)")
    print("4 - Remove water")
    print("5 - Check inside")
    print("6 - Exit")

    local choice = io.read("*n")

    local success,result = pcall(function()
        if choice == 1 then
            print("Enter the amount of rice to put (max 10 kg):")
            local amount = io.read("*n")
            AddRice(amount)
            MainStorageMenu()
        elseif choice == 2 then
            print("Enter the amount of rice to remove:")
            local amount = io.read("*n")
            RemoveRice(amount)
            MainStorageMenu()
        elseif choice == 3 then
            print("Enter the amount of water to put (max 10 L):")
            local amount = io.read("*n")
            AddWater(amount)
            MainStorageMenu()
        elseif choice == 4 then
            print("Enter the amount of water to remove:")
            local amount = io.read("*n")
            RemoveWater(amount)
            MainStorageMenu()
        elseif choice == 5 then
            print("Checking inside the main storage...")
            print("Current amount of rice: " .. RICE_STORAGE .. " kg")
            print("Current amount of water: " .. WATER_STORAGE .. " L")
            MainStorageMenu()
        elseif choice == 6 then
            print("Exiting main storage menu.")
            MainMenu()
        else
            error("Invalid choice. Please try again.")
            MainStorageMenu()
        end
    end)
    if not success then
        print("Error:",result)
        MainStorageMenu()
    end
end

function AddRice(amount)
    local success,result = pcall(function ()
        if((RICE_STORAGE + amount) > 10) then
            error("You are adding too much rice")
        end
        RICE_STORAGE = RICE_STORAGE + amount
        print("Rice added to the storage. Current amount: " .. RICE_STORAGE .. " kg")
    end)
    if not success then
        print("Error:",result)
        MainStorageMenu()
    end
end
function RemoveRice(amount)
    local success,result = pcall(function ()
        if((RICE_STORAGE - amount) > 0) then
            error("You are adding too much rice")
        end
        if((RICE_STORAGE == 0)) then
            error("There is no rice")
        end
        RICE_STORAGE = RICE_STORAGE - amount
        print("Rice removed from the storage. Current amount: " .. RICE_STORAGE .. " kg")
    end)
    if not success then
        print("Error:",result)
        MainStorageMenu()
    end
end

function AddWater(amount)
    local success,result = pcall(function ()
        if((WATER_STORAGE + amount) > 10) then
            error("You are adding too much water")
        end
        WATER_STORAGE = WATER_STORAGE + amount
        print("Water added to the storage. Current amount: " .. WATER_STORAGE .. " L")
    end)
    if not success then
        print("Error:",result)
        MainStorageMenu()
    end
end
function RemoveWater(amount)
    local success,result = pcall(function ()
        if((WATER_STORAGE - amount) > 0) then
            error("You are adding too much water")
        end
        if((WATER_STORAGE == 0)) then
            error("There is no water")
        end
        WATER_STORAGE = WATER_STORAGE - amount
        print("Water removed from the storage. Current amount: " .. WATER_STORAGE .. " L")
    end)
    if not success then
        print("Error:",result)
        MainStorageMenu()
    end
end

function SteamingBasketMenu()
    print("Steaming Basket Options:")
    print("1 - Put ingredients in it")
    print("2 - Remove ingredients")
    print("3 - Check inside")
    print("4 - Exit")

    local choice = io.read("*n")
    local success,result = pcall(function()
        if choice == 1 then
            AddIngredientToBasket()
            SteamingBasketMenu()
        elseif choice == 2 then
            RemoveIngredientFromBasket()
            SteamingBasketMenu()
        elseif choice == 3 then
            print("Checking inside the steaming basket...")
            print("Currently contains ingredients:")
            for _, ingredient in ipairs(ingredients_basket) do
                print("-", ingredient)
            end
            SteamingBasketMenu()
        elseif choice == 4 then
            print("Exiting steaming basket menu.")
            MainMenu()
        else
            error("Invalid choice. Please try again.")
            SteamingBasketMenu()
        end
    end)
    if not success then
        print("Error:",result)
        SteamingBasketMenu()
    end
end

function AddIngredientToBasket()
    local success, result = pcall(function()
        if #ingredients_basket > 10 then
            error("The basket is full")
        end
        print("Choose the type of ingredient to add:")
        print("1 - Vegetables")
        print("2 - Meat")
        print("3 - Others")

        local choice = io.read("*n")
        io.read()

        local ingredientType
        if choice == 1 then
            ingredientType = "Vegetables"
        elseif choice == 2 then
            ingredientType = "Meat"
        elseif choice == 3 then
            ingredientType = "Others"
        else
            error("Invalid choice. Ingredient not added.")
        end

        print("Enter the specific ingredient:")
        local specificIngredient = io.read()

        local fullIngredient = ingredientType .. ": " .. specificIngredient
        table.insert(ingredients_basket, fullIngredient)
        print("Ingredient added to the steaming basket.")
    end)

    if not success then
        print("Error:", result)
    end
end

function RemoveIngredientFromBasket()
    local success, result = pcall(function()
        if #ingredients_basket == 0 then
            error("The basket is empty")
        end
        print("Enter the ingredient to remove:")
        local ingredient = io.read("*l")

        if not ingredient then
            error("Invalid input. Ingredient not removed.")
        end

        ingredient = ingredient:match("^%s*(.-)%s*$")

        if IsIngredientInBasket(ingredient) then
            for i, value in ipairs(ingredients_basket) do
                if value == ingredient then
                    table.remove(ingredients_basket, i)
                    print("Ingredient removed from the steaming basket.")
                    return
                end
            end
        else
            error("Ingredient not found in the steaming basket.")
        end
    end)

    if not success then
        print("Error:", result)
    end
end

function IsIngredientInBasket(ingredient)
    for _, value in ipairs(ingredients_basket) do
        if value == ingredient then
            return true
        end
    end
    return false
end

function SetTIMER()
    local success, result = pcall(function()
        if power_on == false then
            error("Rice cooker is plugged out")
        elseif lid_closed == false then
            error("Rice cooker is open")
        elseif RICE_STORAGE > 0 or #ingredients_basket > 0 then
            if WATER_STORAGE == 0 then
                error("There is no water")
            end
            print("Enter the TIMER duration in minutes:")
            local duration = io.read("*n")
            TIMER = duration
            is_heating = false
            is_under_pressure = true
            for i = duration,0,-1
            do
                print(i)
            end
            print("Finished")
            is_under_pressure = false
            is_warm = true
        else
            error("The rice cooker is empty")
        end
    end)
    if not success then
        print("Error:", result)
    end
end

function StopTIMER()
    TIMER = 0
    print("You stopped the TIMER")
end

MainMenu()

File: rice_cooker.rb
require_relative 'Exception/empty_basket'
require_relative 'Exception/it_is_cooking'
require_relative 'Exception/lid_is_open'
require_relative 'Exception/missing_ingredients'
require_relative 'Exception/no_more_rice'
require_relative 'Exception/no_more_water'
require_relative 'Exception/no_more_space'
require_relative 'Exception/not_plugged_in'
require_relative 'Exception/too_much_rice'
require_relative 'Exception/too_much_water'

class RiceCooker
  def initialize
    @storage_rice = 0
    @storage_water = 0
    @ingredients_basket = []
    @power_on = false
    @lid_closed = true
    @keep_warm_mode = false
    @is_heating = false
    @is_under_pressure = false
    @is_warm = false
    @timer = 0
  end 

  def display_options
    puts "1 - Plug In/Plug Out"
    puts "2 - Open/Close"
    puts "3 - Keep Warm Mode On/Off"
    puts "4 - Check Status Light"
    puts "5 - Put Ingredients in the Main Storage"
    puts "6 - Put Ingredients in the Steaming Basket"
    puts "7 - Set the Timer"
    puts "8 - Stop the Timer"
    puts "9 - Exit"
  end

  def main_menu
    loop do
      display_options
      print "Select an option: "
      option = gets.chomp.to_i

      case option
      when 1
        plug_in_plug_out
      when 2
        open_close
      when 3
        keep_warm_mode
      when 4
        check_status_light
      when 5
        ingredients_storage_menu
      when 6
        steaming_basket_menu
      when 7
        set_timer
      when 8
        stop_timer  
      when 9
        exit_program
      else
        puts "Invalid option. Please try again."
      end
    end
  end

  def plug_in_plug_out
    begin
      raise ITISCOOKING if @timer > 0
      @power_on = !@power_on
      @is_heating = !@is_heating
      puts "Rice cooker is now plugged #{@power_on ? 'in' : 'out'}."
    rescue ITISCOOKING => e
      puts "Error: #{e.message}"
    end 
    
  end

  def open_close
    begin
      raise ITISCOOKING if @timer > 0
      @lid_closed = !@lid_closed
    puts "Rice cooker is now #{@lid_closed ? 'closed' : 'open'}."
    rescue ITISCOOKING => e
      puts "Error: #{e.message}"
    end 
  end

  def keep_warm_mode
    @keep_warm_mode = !@keep_warm_mode
    puts "Keep Warm Mode is now #{@keep_warm_mode ? 'On' : 'Off'}."
  end

  def check_status_light
    if @is_heating == true then
      puts "Rice Cooker is heating now"
    elsif @is_under_pressure == true then
      puts "Rice Cooker is under pressure now"
    elsif @is_warm == true then
      puts "Rice Cooker is now warm" 
    else 
      puts "Rice cooker is off"
    end
  end

  def ingredients_storage_menu
    loop do
      puts "1 - Put Rice in the Storage"
      puts "2 - Remove Rice from the Storage"
      puts "3 - Put Water in the Storage"
      puts "4 - Remove Water from the Storage"
      puts "5 - Check Inside"
      puts "6 - Exit"
      print "Select an option: "
      option = gets.chomp.to_i

      case option
      when 1
        put_rice_in_storage
      when 2
        remove_rice_from_storage
      when 3
        put_water_in_storage
      when 4
        remove_water_from_storage
      when 5
        check_inside
      when 6
        break
      else
        puts "Invalid option. Please try again."
      end
    end
  end

  def put_rice_in_storage
    print "Enter the amount of rice to put in the storage (kg): "
    amount = gets.chomp.to_f
    @storage_rice += amount
    begin
      raise TooMuchRice if @storage_rice > 10
      puts "#{amount} kg of rice is now in the storage."
    rescue TooMuchRice => e
      @storage_rice -= amount
      puts "Error: #{e.message}"
    end
  end

  def remove_rice_from_storage
    print "Enter the amount of rice to remove from the storage (kg): "
    amount = gets.chomp.to_f
    begin
      raise NoMoreRice if amount <= @storage_rice
      @storage_rice -= amount
      puts "#{amount} kg of rice is removed from the storage."
    rescue NoMoreRice => e
      puts "Error: #{e.message}"
    end
  end

  def put_water_in_storage
    print "Enter the amount of water to put in the storage (L): "
    amount = gets.chomp.to_f
    @storage_water += amount
    begin
      raise TooMuchWater if @storage_water > 10
      puts "#{amount} L of water is now in the storage."
    rescue TooMuchWater => e
      @storage_water -= amount
      puts "Error: #{e.message}"
    end
  end

  def remove_water_from_storage
    print "Enter the amount of water to remove from the storage (L): "
    amount = gets.chomp.to_f
    begin
      raise NoMoreWater if amount <= @storage_water
      @storage_water -= amount
      puts "#{amount} L of water is removed from the storage."
    rescue NoMoreWater => e
      puts "Error: #{e.message}"
    end
  end

  def check_inside
    puts "Inside the rice cooker: "
    puts "Rice in Storage: #{@storage_rice} kg"
    puts "Water in Storage: #{@storage_water} L"
    puts "Ingredients in Steaming Basket: #{@ingredients_basket.join(', ')}"
  end

  def steaming_basket_menu
    loop do
      puts "1 - Put Ingredients in it"
      puts "2 - Remove Ingredients"
      puts "3 - Check Inside"
      puts "4 - Exit"
      print "Select an option: "
      option = gets.chomp.to_i

      case option
      when 1
        put_ingredients_in_basket
      when 2
        remove_ingredients_from_basket
      when 3
        check_inside_basket
      when 4
        break
      else
        puts "Invalid option. Please try again."
      end
    end
  end

  def put_ingredients_in_basket
    print "Enter the ingredient to put in the steaming basket: "
    ingredient = gets.chomp
    begin
      raise NoMoreSpace if @ingredients_basket.length() > 3
      @ingredients_basket << ingredient
      puts "#{ingredient} is now in the steaming basket."
    rescue NoMoreSpace => e
      puts "Error: #{e.message}"
    end
  end

  def remove_ingredients_from_basket
    print "Enter the ingredient to remove from the steaming basket: "
    ingredient = gets.chomp
    begin
      raise EmptyBasket if @ingredients_basket.length() == 0
      raise MissingIngredients if @ingredients_basket.include?(ingredient) == false
      @ingredients_basket.delete(ingredient)
      puts "#{ingredient} is removed from the steaming basket."
    rescue EmptyBasket => e
      puts "Error: #{e.message}"
    end
  end

  def check_inside_basket
    puts "Inside the steaming basket: "
    puts "Ingredients: #{@ingredients_basket}"
  end

  def set_timer
    if @storage_rice > 0 || @ingredients_basket.length() > 0 then
      begin
        raise NotPluggedInError if @power_on == false
        raise LidIsOpen if @LidIsOpen == true
        raise NoMoreWater if @storage_water == 0
        print "Enter the timer value (minutes): "
        @timer = gets.chomp.to_i
        puts "Timer is set to #{@timer} minutes."
        @is_heating = true
        countdown_Timer(@timer*60)
      rescue NotPluggedInError => e
        puts "Error: #{e.message}"
      rescue LidIsOpen => e
        puts "Error: #{e.message}"
      rescue NoMoreWater => e
        puts "Error: #{e.message}"  
      end
    end
    puts "There is nothing inside"  
  end

  def stop_timer
    @timer = 0
  end

  def countdown_Timer(seconds)
    @is_heating = false
    @is_under_pressure = true
    Thread.new do
      while seconds > 0
        sleep(1)
        seconds -= 1
      end
    puts "\rTime's up"
    @is_under_pressure = false
    @is_warm = true
    @timer = 0    
    end
  end    

  def exit_program
    puts "Exiting the rice cooker program. Goodbye!"
    exit
  end
end

# Create an instance of the RiceCooker class and start the program
rice_cooker = RiceCooker.new
rice_cooker.main_menu
